mempool/README.md

```diff
@@ -0,0 +1,478 @@
+    
+    // Optional: Custom Cosmos pool (defaults to PriorityNonceMempool)  
+    CosmosPool    sdkmempool.ExtMempool
+    
+    // Optional: Custom broadcast function for promoted transactions
+    BroadCastTxFn func(txs []*ethtypes.Transaction) error
+}
+```
+
+**Minimal Configuration**:
+
+```go
+mempoolConfig := &evmmempool.EVMMempoolConfig{
+    AnteHandler:   app.GetAnteHandler(),
+    BlockGasLimit: 100_000_000, // 100M gas limit
+}
+```
+
+**Custom Cosmos Mempool Configuration**:
+
+The mempool uses a `PriorityNonceMempool` for Cosmos transactions by default. You can customize the priority calculation:
+
+```go
+// Define custom priority calculation for Cosmos transactions
+priorityConfig := sdkmempool.PriorityNonceMempoolConfig[math.Int]{
+    TxPriority: sdkmempool.TxPriority[math.Int]{
+        GetTxPriority: func(goCtx context.Context, tx sdk.Tx) math.Int {
+            feeTx, ok := tx.(sdk.FeeTx)
+            if !ok {
+                return math.ZeroInt()
+            }
+            
+            // Get fee in bond denomination
+            bondDenom := "uatom" // or your chain's bond denom
+            fee := feeTx.GetFee()
+            found, coin := fee.Find(bondDenom)
+            if !found {
+                return math.ZeroInt()
+            }
+            
+            // Calculate gas price: fee_amount / gas_limit
+            gasPrice := coin.Amount.Quo(math.NewIntFromUint64(feeTx.GetGas()))
+            return gasPrice
+        },
+        Compare: func(a, b math.Int) int {
+            return a.BigInt().Cmp(b.BigInt()) // Higher values have priority
+        },
+        MinValue: math.ZeroInt(),
+    },
+}
+
+mempoolConfig := &evmmempool.EVMMempoolConfig{
+    AnteHandler:   app.GetAnteHandler(),
+    BlockGasLimit: 100_000_000,
+    CosmosPool:    sdkmempool.NewPriorityMempool(priorityConfig),
+}
+```
+
+**Custom Block Gas Limit**:
+
+```go
+// Example: 50M gas limit for lower capacity chains
+mempoolConfig := &evmmempool.EVMMempoolConfig{
+    AnteHandler:   app.GetAnteHandler(),
+    BlockGasLimit: 50_000_000,
+}
+```
+
+### Prerequisites
+
+1. **EVM Module Integration**: EVM keeper and module initialized before mempool
+2. **FeeMarket Module**: Required for base fee calculations  
+3. **Compatible AnteHandler**: Must support EVM transaction validation
+
+## Concepts
+
+### Problem Statement
+
+The default CometBFT mempool is incompatible with Ethereum tooling (Forge, Hardhat, deployment scripts) due to fundamental differences in transaction ordering expectations:
+
+1. **FIFO vs Priority Ordering**: CometBFT uses strict FIFO ordering, while Ethereum tools expect fee-based prioritization
+2. **Nonce Gap Rejection**: CometBFT immediately rejects out-of-order nonces, while Ethereum tools expect such transactions to queue until executable
+3. **Base Fee Dynamics**: Transactions may become temporarily invalid due to base fee increases but should remain in the mempool
+
+Example incompatibility from OP Stack deployment:
```

Comment:
could we link to something like a test or a script?

Reply:
fixed: https://github.com/cosmos/evm/pull/462/commits/a5215d284c3dddabe7efd209a0f86f41ce3b18eb